lib/commanded/event/handler.ex

```diff
@@ -933,12 +1080,98 @@ defmodule Commanded.Event.Handler do
             ", expected `:ok` or `{:error, term}`"
         )
 
-        telemetry_stop(start_time, Map.put(telemetry_metadata, :error, :invalid_return_value))
+        telemetry_stop(
+          start_time,
+          Map.put(telemetry_metadata, :error, :invalid_return_value),
+          :handle
+        )
 
         error = {:error, :invalid_return_value}
         failure_context = build_failure_context(event, context, state)
 
-        handle_event_error(error, event, failure_context, state)
+        next = fn context, state -> handle_event(event, context, state) end
+        handle_event_error(error, event, failure_context, state, next)
+    end
+  end
+
+  defp handle_batch(events, context \\ %{}, handler)
+
+  defp handle_batch(events, context, %Handler{last_seen_event: last_seen_event} = state)
+       when is_number(last_seen_event) do
+    %{event_number: last_event_number} = last_event = List.last(events)
+
+    if last_event_number <= last_seen_event do
+      Logger.debug(fn ->
+        describe(state) <> " has already seen event ##{inspect(last_event_number)}"
+      end)
+
+      confirm_receipt([last_event], state)
+    else
+      events
+      |> Enum.reject(&(&1.event_number <= last_seen_event))
+      |> do_handle_batch(context, state)
+    end
+  end
+
+  defp handle_batch(events, context, %Handler{} = state) do
+    do_handle_batch(events, context, state)
+  end
+
+  defp do_handle_batch([], _context, state), do: state
+
+  defp do_handle_batch(events, context, %Handler{} = state) do
+    %Handler{handler_module: handler_module} = state
+
+    telemetry_metadata = batch_telemetry_metadata(events, context, state)
+    start_time = telemetry_start(telemetry_metadata, :batch)
+
+    enriched_events =
+      Enum.map(events, fn e = %RecordedEvent{data: data} ->
+        {data, enrich_metadata(e, state)}
+      end)
+
+    try do
+      case handler_module.handle_batch(enriched_events) do
+        :ok ->
+          telemetry_stop(start_time, telemetry_metadata, :batch)
+          confirm_receipt(events, state)
```

Comment:
We pass `enriched_events` to the handler callback, but the unenriched list for error handling. Would it make sense to use the enriched_events here and below?

Reply:
I think I'm going to keep this as it is for now since this is also how handle_events behaves